### PR TITLE
define Suspicion model

### LINKED ISSUE #41: [suspicion] suspicion model
define suspicion model. This struct manages suspect timer

* each suspect has its own timer
* timer could be accelerate according to the Lifeguard principle "Dynamic suspicion timeouts"

### PR DESCRIPTION
**Pull request by zeroFruit**
resolved: #41 

details:

image

(image from hashicorp)

for dynamically determine suspect timeout created timer for suspect member.

### CODE REVIEW

**Comment by junbeomlee** [comment]
Is the suspicion structure that manages the suspected members?

**Comment by zeroFruit** [comment]
no, the object who use this `Suspicion` struct would manage members

so I think the structure would be like this:

```go
struct Memberlist {
    members []Member
    suspicions []Suspicion
}
```

File: suspicion.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+// TODO: This returns true if this confirmer is new, and false if it was a duplicate information
+// TODO: or if we've got enough confirmations to hit the value of timer to minimum
+func (s *Suspicion) Confirm(confirmer MemberID) bool {
+	return false
+}
+
+// TODO: remainingSuspicionTime helps to calculate the remaining time to wait before suspected node
+// TODO: considered as a dead. The return value could be negative, in the case of return value
+// TODO: is negative, immediately fire the timer
+func remainingSuspicionTime(n, k uint32, elapsed, min, max time.Duration) time.Duration {
```
Comment (hihiboss) [inline suspicion.go]: It would be better that the function name start with a verb.

**Comment by hihiboss** [comment]
I think `Suspicion` should be used in `status` of `Member`. Because `Memeber` has `status` and this could be Unknown, alive, **suspected**, dead.

**Comment by zeroFruit** [comment]
> I think Suspicion should be used in status of Member. Because Memeber has status and this could be Unknown, alive, suspected, dead.

But the suspicion is more related with timer not status.
I've commented on top of `Suspicion` struct. What `Suspicion` mainly do is calculate left Suspicion time, and manage suspicion time for each suspected member
```
// suspicion manages the suspect timer and helps to accelerate the timeout
``

**Comment by junbeomlee** [comment]
I think we should talk more about this part.
Whether to suspicion and member management at the same time, or not

**Comment by junbeomlee** [comment]
We decided to move the suspicion into the member.

**Pull request closed without merging**